File: instrumentation/network/src/main/java/io/opentelemetry/android/instrumentation/network/NetworkChangeInstrumentation.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.network;

import androidx.annotation.NonNull;
import com.google.auto.service.AutoService;
import io.opentelemetry.android.common.internal.features.networkattributes.data.CurrentNetwork;
import io.opentelemetry.android.instrumentation.AndroidInstrumentation;
import io.opentelemetry.android.instrumentation.InstallationContext;
import io.opentelemetry.android.internal.services.Services;
import io.opentelemetry.api.common.AttributesBuilder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.BiConsumer;

/** Generates telemetry for when the network status changes. */
@AutoService(AndroidInstrumentation.class)
public final class NetworkChangeInstrumentation implements AndroidInstrumentation {

    final List<BiConsumer<AttributesBuilder, CurrentNetwork>> additionalExtractors =
            new ArrayList<>();

    /** Adds a {@link BiConsumer} that can add Attributes about the current network. */
    public NetworkChangeInstrumentation addAttributesExtractor(
            BiConsumer<AttributesBuilder, CurrentNetwork> extractor) {
        additionalExtractors.add(extractor);
        return this;
    }

    @Override
    public void install(@NonNull InstallationContext ctx) {
        additionalExtractors.add(new NetworkChangeAttributesExtractor());
        Services services = Services.get(ctx.getApplication());
        NetworkChangeMonitor networkChangeMonitor =
                new NetworkChangeMonitor(
                        ctx.getOpenTelemetry(),
                        services.getAppLifecycle(),
                        services.getCurrentNetworkProvider(),
                        Collections.unmodifiableList(additionalExtractors));
        networkChangeMonitor.start();
    }
}

File: instrumentation/network/src/main/java/io/opentelemetry/android/instrumentation/network/NetworkChangeInstrumentation.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.network

import android.content.Context
import com.google.auto.service.AutoService
import io.opentelemetry.android.OpenTelemetryRum
import io.opentelemetry.android.common.internal.features.networkattributes.data.CurrentNetwork
import io.opentelemetry.android.instrumentation.AndroidInstrumentation
import io.opentelemetry.android.internal.services.Services.Companion.get
import io.opentelemetry.api.common.AttributesBuilder

/**
 * A tag interface for an extractor that can add attributes from the [CurrentNetwork].
 */
fun interface NetworkAttributesExtractor : (AttributesBuilder, CurrentNetwork) -> Unit

/** Generates telemetry for when the network status changes.  */
@AutoService(AndroidInstrumentation::class)
class NetworkChangeInstrumentation : AndroidInstrumentation {
    private val additionalAttributeExtractors: MutableList<NetworkAttributesExtractor> = ArrayList()
    private var networkApplicationListener: NetworkApplicationListener? = null

    override val name = "network"

    /** Adds a [NetworkAttributesExtractor] that can add Attributes from the [CurrentNetwork].  */
    fun addAttributesExtractor(attributeExtractor: NetworkAttributesExtractor): NetworkChangeInstrumentation {
        additionalAttributeExtractors.add(attributeExtractor)
        return this
    }

    override fun install(context: Context, openTelemetryRum: OpenTelemetryRum) {
        additionalAttributeExtractors.add(NetworkChangeAttributesExtractor())
        val services = get(context)
        val listener = NetworkApplicationListener(services.currentNetworkProvider)
        val logger = openTelemetryRum.openTelemetry.logsBridge["io.opentelemetry.network"]
        listener.startMonitoring(logger, additionalAttributeExtractors)
        services.appLifecycle.registerListener(listener)
        networkApplicationListener = listener
    }

    override fun uninstall(context: Context, openTelemetryRum: OpenTelemetryRum) {
        networkApplicationListener?.let { listener ->
            listener.stopMonitoring()
            val services = get(context)
            services.appLifecycle.unregisterListener(listener)
        }
        networkApplicationListener = null
    }
}
